### PR TITLE
[1.12] Allow enchantments to modify damage based on the entity hit

### PR DESCRIPTION
**Pull request by maxanier**
## Why
Enchantments are currently able to make enchanted weapons deal more damaged based on the hit entity's `EnumCreatureAttribute`.
However, it might be interesting to change this based on the entity's attributes. For example I would like to create an enchantment that only deals extra damage do vampire players, therefore I need to check the player entity's capabilities. 
I've tried to use the `Enchantment#onEntityDamaged` (attack the player again in there) for this, but this does not really work as some of the damage calculation is ignored then. Also it is called twice after attacking somehow.

## What
This PR adds an additional method (`calcDamageByEntity`), which calls the 'old' (`calcDamageByCreature`) by default, to the Enchantment class.
`EnchantmentHelper` calls the new method instead of the old one now.
Unfortunately a few more classes had be modified to pass the `Entity` parameter instead of the `EnumCreatureAttribute`.

The entity parameter can be null (ItemStacks use this to generate a damage value for the description). I've added a `Nullable` annotation to make this clear.

## Addendum
I was not sure if I should deprecate the `calcDamageByCreature` method as it might still be interesting for modders to override, but should not be called. (While writing this I decided that it would probably be a good idea)

Also not sure about commenting in patches, so I only added ones where they are necessary.

### CODE REVIEW

File: patches/minecraft/net/minecraft/enchantment/Enchantment.java.patch
Comment (liach) [inline patches/minecraft/net/minecraft/enchantment/Enchantment.java.patch]: Should probably deprecate this one instead.

File: patches/minecraft/net/minecraft/enchantment/Enchantment.java.patch
Comment (liach) [inline patches/minecraft/net/minecraft/enchantment/Enchantment.java.patch]: You should move this to a forge block in the end of the class.

File: patches/minecraft/net/minecraft/enchantment/EnchantmentHelper.java.patch
Comment (liach) [inline patches/minecraft/net/minecraft/enchantment/EnchantmentHelper.java.patch]: Probably leaving the old one and deprecating it is better.

File: patches/minecraft/net/minecraft/enchantment/EnchantmentHelper.java.patch
Comment (liach) [inline patches/minecraft/net/minecraft/enchantment/EnchantmentHelper.java.patch]: Call this entity... Or there may be confusions for mappings.

File: patches/minecraft/net/minecraft/enchantment/Enchantment.java.patch
Comment (maxanier) [reply patches/minecraft/net/minecraft/enchantment/Enchantment.java.patch]: Wanted to keep the patch small, but does make stuff cleaner. Done.

File: patches/minecraft/net/minecraft/enchantment/EnchantmentHelper.java.patch
Comment (maxanier) [reply patches/minecraft/net/minecraft/enchantment/EnchantmentHelper.java.patch]: Should the new method be moved to a Forge block or stay next to this? :D

Comment (liach) [reply patches/minecraft/net/minecraft/enchantment/EnchantmentHelper.java.patch]: Probably leave it before subclasses and after all methods (in a forge block is better)

**Comment by maxanier** [comment]
Updated based on liach's suggestions (thx).

**Review by liach — APPROVED** [review]

**Comment by liach** [comment]
This looks great to me.

**Comment by maxanier** [comment]
@mezz Any concerns or is this good to go?

File: patches/minecraft/net/minecraft/enchantment/EnchantmentHelper.java.patch
Comment (mezz) [inline patches/minecraft/net/minecraft/enchantment/EnchantmentHelper.java.patch]: I'm not sure I see the purpose of changing this.
Deprecating it is fine, but why return 0?

**Review by mezz — CHANGES_REQUESTED** [review]

File: patches/minecraft/net/minecraft/enchantment/EnchantmentHelper.java.patch
Comment (maxanier) [reply patches/minecraft/net/minecraft/enchantment/EnchantmentHelper.java.patch]: Unobfuscated old code:
``` java
        ENCHANTMENT_MODIFIER_LIVING.livingModifier = 0.0F;
        ENCHANTMENT_MODIFIER_LIVING.entityLiving = creatureAttribute;
        applyEnchantmentModifier(ENCHANTMENT_MODIFIER_LIVING, stack);
        return ENCHANTMENT_MODIFIER_LIVING.livingModifier;
```

The old code is not working anymore since the ENCHANTMENT_MODIFIER_LIVING now requires an Entity instead of an EnumCreatureAttribute to calculate it's value. Since this is not available in this context I decided to just return 0 in this deprecated method.

This might actually cause minor problems with mods that do use this method directly instead of indirectly via EntityPlayer or EntityMob as those mods would not find out the actual value for some enchantments.

However, the only other (non breaking) solution I can think of, is to create second ENCHANTMENT_MODIFIER_(CREATURE_OLD) (or integrate it in the current one) which can make use of a EntityCreatureAttribute. But since this would ignore any values added to the enchantments in newer ways this would be a somewhat ugly solution. It would solve most of the issues through.

*EDIT:* After writing this, I think, using the ugly solution might be better than not using it.
The issue of "old" mods not getting the correct value for all enchantments is not that problematic, but
without a breaking change it is kinda hard to notice. 
So this would allow old mods to get the correct value for the old type (includes vanilla) enchantments at least.

 *Usage*     | *Enchantment*   | *Works*              |
|-------------|-----------------|----------------------|
| Vanilla     | Old and New     | Yes                  |
| Updated Mod | Old and New     | Yes                  |
| Old Mod     | Old and Vanilla | Only with workaround |
| Old Mod     | New             | No                   |

Comment (mezz) [reply patches/minecraft/net/minecraft/enchantment/EnchantmentHelper.java.patch]: I think we should try to support old mods without breaking them.

**Comment by maxanier** [comment]
Rebased to resolve conflicts and adjusted so that if mods call the old method, so only an EnumEntityAttribute is available, the old Enchantment#calcDamageByCreature is used. Which should get the correct result for vanilla and any mods using the old method.

File: patches/minecraft/net/minecraft/enchantment/Enchantment.java.patch
Comment (liach) [inline patches/minecraft/net/minecraft/enchantment/Enchantment.java.patch]: Can you move 
````patch
+    /**
+     * This applies specifically to applying at the enchanting table. The other method {@link #canApply(ItemStack)}
+     * applies for <i>all possible</i> enchantments.
+     * @param stack
+     * @return
+     */
+    public boolean canApplyAtEnchantingTable(ItemStack stack)
+    {
+        return stack.func_77973_b().canApplyAtEnchantingTable(stack, this);
+    }
+
+    /**
+     * Is this enchantment allowed to be enchanted on books via Enchantment Table
+     * @return false to disable the vanilla feature
+     */
+    public boolean isAllowedOnBooks()
+    {
+        return true;
+    }
````
to the forge block?

File: patches/minecraft/net/minecraft/enchantment/Enchantment.java.patch
Comment (howtonotwin) [inline patches/minecraft/net/minecraft/enchantment/Enchantment.java.patch]: Do __*not*__ add Javadoc to Vanilla methods. Their Javadoc is part of the MCP mappings and should not be modified by Forge. The convention is to just stick a normal comment after `@Deprecated`.

File: patches/minecraft/net/minecraft/enchantment/EnchantmentHelper.java.patch
Comment (howtonotwin) [inline patches/minecraft/net/minecraft/enchantment/EnchantmentHelper.java.patch]: Same as above: do __*not*__ modify Vanilla methods' Javadoc.

File: patches/minecraft/net/minecraft/enchantment/Enchantment.java.patch
Comment (howtonotwin) [inline patches/minecraft/net/minecraft/enchantment/Enchantment.java.patch]: Add a period and

```
@param level the level of this enchantment on the item doing the attack
@param entity the entity being attacked
@return the additional damage to do
```

to the Javadoc (preserving case exactly, so as to conform to the convention).

File: patches/minecraft/net/minecraft/enchantment/EnchantmentHelper.java.patch
Comment (howtonotwin) [inline patches/minecraft/net/minecraft/enchantment/EnchantmentHelper.java.patch]: Javadoc this.

File: patches/minecraft/net/minecraft/enchantment/EnchantmentHelper.java.patch
Comment (howtonotwin) [inline patches/minecraft/net/minecraft/enchantment/EnchantmentHelper.java.patch]: Ternaries should be spaced out, as `cond ? tbr : fbr`, not `cond?tbr:fbr`.

File: patches/minecraft/net/minecraft/entity/player/EntityPlayer.java.patch
Comment (howtonotwin) [inline patches/minecraft/net/minecraft/entity/player/EntityPlayer.java.patch]: Unnecessary patching. The next line sets `f1` too, so there's no need to remove this `if`/`else`.

File: patches/minecraft/net/minecraft/enchantment/Enchantment.java.patch
Comment (howtonotwin) [inline patches/minecraft/net/minecraft/enchantment/Enchantment.java.patch]: Ternaries should be spaced out, as `cond ? tbr : fbr`, not `cond?tbr:fbr`.

File: patches/minecraft/net/minecraft/enchantment/Enchantment.java.patch
Comment (howtonotwin) [inline patches/minecraft/net/minecraft/enchantment/Enchantment.java.patch]: You're missing a space at the end of this. Should be like:

    /* ============ FORGE START ============ */

not

    /* ============ FORGE START ============*/

**Review by howtonotwin — CHANGES_REQUESTED** [review]

File: patches/minecraft/net/minecraft/entity/player/EntityPlayer.java.patch
Comment (maxanier) [reply patches/minecraft/net/minecraft/entity/player/EntityPlayer.java.patch]: sounds strange to do some unnecessary calculations, but it decreases patch size, so ok.

Comment (covers1624) [reply patches/minecraft/net/minecraft/entity/player/EntityPlayer.java.patch]: Or just whack an if (false) at the top of the if statement.

**Comment by maxanier** [comment]
Thx for the feedback. Applied your suggestions

File: patches/minecraft/net/minecraft/enchantment/Enchantment.java.patch
Comment (howtonotwin) [inline patches/minecraft/net/minecraft/enchantment/Enchantment.java.patch]: Typo?

**Review by howtonotwin — CHANGES_REQUESTED** [review]

File: patches/minecraft/net/minecraft/enchantment/Enchantment.java.patch
Comment (maxanier) [reply patches/minecraft/net/minecraft/enchantment/Enchantment.java.patch]: Have missed that when moving the other methods in the Forge block. Thx

**Review by howtonotwin — APPROVED** [review]

**Comment by maxanier** [comment]
Not sure why the licence/cla check fails as I have signed it before (and again now).

But besides that, any other concerns @mezz?
"Old" mods can now get the correct enchantment modifier for all enchantments except modded ones which utilize the new method.

**Comment by mezz** [comment]
Bump, unstale. 
Looks like the CLA check is passing now.
I'll take a look at this soon.

**Comment by mezz** [comment]
Looking good. Please add a test mod so I can try this out, and also to serve as an example usage of this new feature.

**Comment by maxanier** [comment]
Rebased onto latest 1.12.x branch to make sure it still works
Fixed another issue
Added Testmod

From EnchantmentDamageTest:
```
/**
 * Test for calculating enchantment damage modifier based on the entity object instead of the creature attribute.
 * <p>
 * HowTo Test:
 * <p>
 * a) Grab yourself a wooden sword and enchant it with "enchantmentdamagetest.new_enchantment"
 * Spawn a Zombie and any other animal (e.g. a cow)
 * The Zombie should be one hit with this sword, the cow should take multiple hits (as normal)
 * <p>
 * b) To verify that the old (deprecated/vanilla) way/method still works
 * enchant a wooden sword with "enchantmentdamagetest.old_enchantment"
 * and repeat the test with a undead creature (Zombie) and a non undead creature (cow).
 * It should have the same result
 */
```

**Comment by mezz** [comment]
Tested and looks good to me.

**Review by mezz — APPROVED** [review]

File: patches/minecraft/net/minecraft/enchantment/EnchantmentHelper.java.patch
Comment (LexManos) [inline patches/minecraft/net/minecraft/enchantment/EnchantmentHelper.java.patch]: Binary breaking change, we can't allow this...

**Comment by maxanier** [comment]
Mhh, good point, not sure why I changed the name of the attribute anyway, has been some time.

Maybe because I wanted to add a @Nullable annotation and wasn't sure how this would affect de/obfuscation.

I will rename the attribute to the old name. Should I leave the annotation or remove it as well?

**Comment by mezz** [comment]
The name change is in a package-private class so nothing can really reference it or break when it changes.

**Comment by LexManos** [comment]
You know better then that Mezz, modders access all kinda of dumb things.

**Comment by mezz** [comment]
@maxanier you can leave the annotation, it is fine. Please update this to resolve merge conflicts as well.

**Comment by maxanier** [comment]
Rebased and resolved conflicts as well as renamed the attribute to the original name to avoid binary breaking changes.

File: patches/minecraft/net/minecraft/enchantment/EnchantmentHelper.java.patch
Comment (mezz) [inline patches/minecraft/net/minecraft/enchantment/EnchantmentHelper.java.patch]: `Nullable` should go on a new line before the `EnumCreatureAttribute` field, so the original line is not edited.
the new `entity` field should go after the vanilla fields

File: patches/minecraft/net/minecraft/enchantment/EnchantmentHelper.java.patch
Comment (mezz) [inline patches/minecraft/net/minecraft/enchantment/EnchantmentHelper.java.patch]: This newline should not be added

File: patches/minecraft/net/minecraft/entity/player/EntityPlayer.java.patch
Comment (mezz) [inline patches/minecraft/net/minecraft/entity/player/EntityPlayer.java.patch]: leave this newline here, do not remove it

**Review by mezz — CHANGES_REQUESTED** [review]
looks good, just needs some minor fixes for the patches

File: patches/minecraft/net/minecraft/enchantment/EnchantmentHelper.java.patch
Comment (liach) [reply patches/minecraft/net/minecraft/enchantment/EnchantmentHelper.java.patch]: Hmm, this field should be `EnumCreatureAttribute.UNDEFINED` by default instead of `null`.

Comment (mezz) [reply patches/minecraft/net/minecraft/enchantment/EnchantmentHelper.java.patch]: That's how mojang has it (null)... the way the class itself is used is very strange.

File: patches/minecraft/net/minecraft/entity/player/EntityPlayer.java.patch
Comment (maxanier) [reply patches/minecraft/net/minecraft/entity/player/EntityPlayer.java.patch]: ~~Why? Idea was to prevent an additional line as described in the Contribution Guidelines https://github.com/MinecraftForge/MinecraftForge/wiki/If-you-want-to-contribute-to-Forge#conventions-for-coding-patches-for-a-minecraft-class-javapatch~~

File: patches/minecraft/net/minecraft/enchantment/EnchantmentHelper.java.patch
Comment (maxanier) [reply patches/minecraft/net/minecraft/enchantment/EnchantmentHelper.java.patch]: Will change that.

Yeah, this has one of the strangest concepts I have seen in Minecraft so far

File: patches/minecraft/net/minecraft/enchantment/EnchantmentHelper.java.patch
Comment (maxanier) [reply patches/minecraft/net/minecraft/enchantment/EnchantmentHelper.java.patch]: Sorry, will remove that
My git diff highlighting is always confused by all the +- in the patch files :angry:

File: patches/minecraft/net/minecraft/entity/player/EntityPlayer.java.patch
Comment (liach) [reply patches/minecraft/net/minecraft/entity/player/EntityPlayer.java.patch]: that's to avoid patches, but here you are making patches.

Comment (maxanier) [reply patches/minecraft/net/minecraft/entity/player/EntityPlayer.java.patch]: I'm sorry, I don't get this. Aren't you always making patches if you modify/patch Minecraft code?
Fixed anyway

**Comment by maxanier** [comment]
Applied requested changes

File: patches/minecraft/net/minecraft/entity/player/EntityPlayer.java.patch
Comment (maxanier) [reply patches/minecraft/net/minecraft/entity/player/EntityPlayer.java.patch]: Never mind, I think I have misinterpreted the guidelines, sorry

**Review by mezz — APPROVED** [review]

**Comment by LexManos** [comment]
Gunna need to talk to Grum about the lack of thread safety in EnchantmentHelper.... 
Why are you hijacking ENCHANTMENT_MODIFIER_LIVING? Why not just make your own modifier? Then you wouldn't have to hack all the dual fields stuff...
Also your modification to EntityPlayer is just... wrong... At first glance it looks like you're doubling up the modifier. Which you are in a sense. But you're also applying it to the wrong value. 
You're applying the value to the draw damage. Not the enchanted value that is then multiplied by the cooldown strength. You bypass that whole mechanic....

**Comment by maxanier** [comment]
Thanks for noticing the wrong calculation. Sorry about that. I looks like I somehow accidentally changed the `f1` to `f`. Not sure how I was able to screw that up as it was just copy and paste.

Did not create my own modifier because I wanted to keep the changes as minimal as possible (and in one of the earlier version I just replaced the entityAttribute attribute with the entity object as far as I can remember so there was less reason to create my own version.

**Comment by maxanier** [comment]
@LexManos did you have a chance to talk with Grum about the thread safety?

**Comment by maxanier** [comment]
Any news about this? Kind of stale 

Will fix the conflicts once Lex decides if it will be merged (in the current implementation)

**Comment by mezz** [comment]
Waiting on reply from @LexManos

**Comment by LexManos** [comment]
I haven't talked to Grum about the thread safety beyond preliminary 'things are broken'. So there really isn't anything we can do for it for now.

**Comment by liach** [comment]
There is a few merge conflicts.

**Comment by maxanier** [comment]
I won't update this before there is some new information about thread safety from Lex or Grum

**Comment by MajorTuvok** [comment]
This just needs some thread saftey comments...
Go away stale, this is a nice PR...

**Comment by liach** [comment]
as long as entities are read but not modified (which happens most of the time), things should be fine

**Pull request closed without merging**
